android/app/src/main/java/org/uwaterloo/subletr/pages/profile/ProfilePageViewModel.kt

```diff
@@ -0,0 +1,217 @@
+import kotlin.jvm.optionals.getOrNull
+
+@HiltViewModel
+class ProfilePageViewModel @Inject constructor(
+	private val userApi: UserApi,
+	private val listingsApi: ListingsApi,
+	savedStateHandle: SavedStateHandle,
+	private val authenticationService: IAuthenticationService,
+	val navigationService: INavigationService,
+): SubletrViewModel<ProfilePageUiState>() {
+
+	val navHostController: NavHostController get() = navigationService.navHostController
+
+	private val userId: Int = checkNotNull(savedStateHandle["userId"])
+	private val userIdStream: BehaviorSubject<Int> = BehaviorSubject.createDefault(userId)
+	private val nameStream: BehaviorSubject<String> = BehaviorSubject.createDefault("")
+	val listingIdStream: BehaviorSubject<Optional<Int>> =
+		BehaviorSubject.createDefault(Optional.empty())
+	private val listingDetailsStream: BehaviorSubject<Optional<ListingDetails>> =
+		BehaviorSubject.createDefault(Optional.empty())
+	private val listingImageStream: BehaviorSubject<Optional<String>> =
+		BehaviorSubject.createDefault(Optional.empty())
+	val ratingStream: BehaviorSubject<Float> = BehaviorSubject.createDefault(0.0f)
+	val usersRatingStream: BehaviorSubject<Int> = BehaviorSubject.createDefault(0)
+
+	private val userDetailsStream: Observable<Result<GetUserResponse>> = userIdStream.map {
+		runBlocking {
+			runCatching {
+				userApi.userGet(it)
+			}.onSuccess { userResponse ->
+				nameStream.onNext("${userResponse.firstName} ${userResponse.lastName}")
+				ratingStream.onNext(userResponse.rating)
+				if (userResponse.listingId != null) {
+					listingIdStream.onNext(Optional.of(userResponse.listingId))
+					runCatching {
+						listingsApi.listingsDetails(
+							listingId = userResponse.listingId,
+							longitude = UWATERLOO_LONGITUDE,
+							latitude = UWATERLOO_LATITUDE,
+						)
+					}
+						.onSuccess { listing ->
+							listingDetailsStream.onNext(Optional.of(listing.details))
+							if (listing.details.imgIds.isNotEmpty()) {
+								runCatching {
+									listingsApi.listingsImagesGet(
+										listing.details.imgIds.first(),
+									)
+								}
+									.onSuccess { imageResponse ->
+										listingImageStream.onNext(Optional.of(imageResponse))
+									}
+									.onFailure {
+										Log.d("API ERROR", "Failed to get user's listing's image")
+									}
+							}
+						}
+						.onFailure {
+							Log.d("API ERROR", "Failed to get user's listing details")
+						}
+				}
+			}
+		}
+			.onFailure {
+				Log.d("API ERROR", "Failed to get user details")
+				navigationService.navHostController.popBackStack()
+			}
+	}
+		.onErrorResumeWith(Observable.never())
+		.subscribeOn(Schedulers.io())
+
+	private val listingImageBitmapStream: Observable<Optional<Bitmap>> = listingImageStream.map {
+		val nullableVersion = it.getOrNull()
+		if (nullableVersion != null) {
+			Optional.of(nullableVersion.base64ToBitmap())
+		}
+		else {
+			Optional.empty<Bitmap>()
+		}
+	}
+		.observeOn(Schedulers.computation())
```

Comment:
You shouldn't add a `.observeOn` at the end of a stream.
`observeOn` sets the thread pool after it is run so you should do the `.observeOn(Schedulers.computation())` right before you have heavy computations.